README.md
# etecsa-cli

[![npm version](https://img.shields.io/npm/v/etecsa-cli.svg)](https://www.npmjs.com/package/etecsa-cli)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)

ETECSA login manager

## Requirements

- [node and npm](https://nodejs.org/en/)

## Install

```sh
npm i -g etecsa-cli
```

## Usage

```sh
$ etecsa command

Usage

  Command one of

  - set user pass
  - login
  - logout
  - status
  - time

Example

  $ etecsa set myuser mypass
  $ etecsa login
  $ etecsa status
  Connected
  $ etecsa time
  00:34:48
```

Config is stored in ~/.cache/.etecsa.json

## Contributing

```sh
git clone https://github.com/danielpza/etecsa-cli.git
cd etecsa-cli
npm link
```

## Gnome Integration

[Etecsa Login Gnome Shell Extension](https://github.com/danielpza/etecsa-login-gnome-shell-extension)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.3](https://github.com/danielpza/etecsa-cli/compare/v0.1.2...v0.1.3) (2020-01-17)


### Features

* use cache instead ([8dca2e1](https://github.com/danielpza/etecsa-cli/commit/8dca2e1a3343c934a7e5c75f37300821b26700b0))
* use XDG_CONFIG_HOME ([daaabb9](https://github.com/danielpza/etecsa-cli/commit/daaabb9fa8637d9894a00abc56bf071d045674ac))


### Bug Fixes

* hotfix broken build ([f7db085](https://github.com/danielpza/etecsa-cli/commit/f7db085a0c38812654f5b0f8fc7adc3eb00dc449))

### [0.1.2](https://github.com/danielpza/etecsa-cli/compare/v0.1.1...v0.1.2) (2019-05-29)


### Bug Fixes

* **status:** report not connected if got fails ([9a9b7fe](https://github.com/danielpza/etecsa-cli/commit/9a9b7fe))


src/core.ts
import got from "got";
import cheerio from "cheerio";

import { CookieJar } from "tough-cookie";

const cookieJar = new CookieJar();

const GOOGLE = "http://www.google.com";
const ETECSA_LOGIN = "https://secure.etecsa.net:8443";
const ETECSA_LOGIN_POST = "https://secure.etecsa.net:8443//LoginServlet";
const ETECSA_LOGOUT = "https://secure.etecsa.net:8443/LogoutServlet";
const ETECSA_TIME = "https://secure.etecsa.net:8443/EtecsaQueryServlet";

export async function login(username: string, password: string) {
  const params = await getLoginParameters();
  const loginParams = { ...params, username, password };
  const loginResponse = await postLogin(loginParams);
  const uuid = await getUuid(loginResponse);
  return uuid;
}

export async function logout(ATTRIBUTE_UUID: string) {
  const { body } = await got.get(ETECSA_LOGOUT, {
    cookieJar,
    query: {
      ATTRIBUTE_UUID,
      remove: "1"
    }
  });
  if (!body.match(/success/i)) {
    throw new Error("Error disconnecting");
  }
}

export async function status() {
  try {
    const { body } = await got.get(GOOGLE);
    return !body.match(ETECSA_LOGIN);
  } catch {
    return false;
  }
}

interface LoginParameters {
  wlanuserip: string;
  wlanacname: string;
  wlanmac: string;
  firsturl: string;
  ssid: string;
  usertype: string;
  gotopage: string;
  successpage: string;
  loggerId: string;
  lang: string;
  username: string;
  password: string;
  CSRFHW: string;
}

async function getLoginParameters(): Promise<LoginParameters> {
  const loginResult = await got.get(ETECSA_LOGIN, { cookieJar });

  const $ = cheerio.load(loginResult.body);
  const result = $("form#formulario")
    .find("input")
    .filter((_, el) => ["button", "reset"].indexOf($(el).attr("type")!) === -1)
    .map((_, el) => ({ name: $(el).attr("name"), value: $(el).val() }))
    .get() as { name: string; value: string }[];
  const map = {} as any;
  for (const { name, value } of result) {
    map[name] = value;
  }
  return map;
}

async function postLogin(parameters: object) {
  const { body, statusCode, statusMessage, headers } = await got.post(
    ETECSA_LOGIN_POST,
    {
      form: true,
      followRedirect: true,
      throwHttpErrors: false,
      cookieJar,
      body: parameters
    }
  );
  const alertMatch = body.match(
    /<script type="text\/javascript">\s*alert\("([\w ().\p{L}]+)"\);/u
  );
  if (alertMatch) {
    throw new Error(alertMatch[1]);
  }
  if (statusCode !== 302) {
    throw new Error(`Server responded with ${statusCode}: ${statusMessage}`);
  }
  return headers.location!; // TODO check for location to be undefined
}

async function getUuid(url: string) {
  const { body } = await got.get(url, { cookieJar });
  const uuidMatch = body.match(/ATTRIBUTE_UUID=(\w*)&/);
  if (!uuidMatch) {
    throw new Error("No uuid in page");
  }
  return uuidMatch[1];
}

export async function getTime(username: string) {
  const { body } = await got.post(ETECSA_TIME, {
    cookieJar,
    query: { op: "getLeftTime", op1: username }
  });
  return body;
}


src/index.ts
import * as api from "./core";
import { resolve } from "path";
import { readFile as readFileCb, writeFile as writeFileCb } from "fs";
import { promisify } from "util";
import { homedir } from "os";

const readFile = promisify(readFileCb);
const writeFile = promisify(writeFileCb);

const configFile = resolve(
  process.env.XDG_CACHE_HOME || resolve(homedir(), ".cache"),
  "etecsa.json"
);

export async function login() {
  const { user, pass } = await readConfig();
  const uuid = await api.login(user, pass);
  console.log(`Logged in`);
  await writeConfig({ user, pass, uuid });
}

export async function logout() {
  const { user, pass, uuid } = await readConfig();
  await api.logout(uuid);
  console.log("Logged out");
  await writeConfig({ user, pass });
}

export async function set(user: string, pass: string) {
  await writeConfig({ user, pass });
}

export async function status() {
  if (await api.status()) {
    console.log("Connected");
  } else {
    console.log("Not Connected");
  }
}

export async function time() {
  const { user } = await readConfig();
  console.log(await api.getTime(user));
}

async function readConfig() {
  const file = await readFile(configFile);
  return JSON.parse(file.toString());
}

async function writeConfig(config: any) {
  await writeFile(configFile, JSON.stringify(config));
}


src/cli.ts
#!/usr/bin/env node
import { login, logout, set, status, time } from "./index";

const option = process.argv[2];

if (!option) {
  console.error("Must select option");
  showHelp();
  process.exit(1);
}

if (option === "set") {
  const [user, pass] = process.argv.slice(3);
  set(user, pass).catch(handleError);
} else if (option === "login") {
  login().catch(handleError);
} else if (option === "logout") {
  logout().catch(handleError);
} else if (option === "status") {
  status().catch(handleError);
} else if (option === "time") {
  time().catch(handleError);
} else {
  console.error(`unknown command ${option}`);
  showHelp();
  process.exit(1);
}

function handleError(err: Error) {
  console.error(err);
  process.exit(1);
}

function showHelp() {
  console.log(`\
$ etecsa command

Usage

  Command one of

  - set user pass
  - login
  - logout
  - status
  - time

Example

  $ etecsa set myuser mypass
  $ etecsa login
  $ etecsa status
  Connected
  $ etecsa time
  00:34:48
`);
}
